File: tests/gastest/test_components.js
/**
 * RUN EVERY COMPONENT'S SCRIPT AGAINST ITS OWN MARKUP.
 *
 * Reading the source was not enough, and I know that because I wrote a
 * reading test first and its mutation SURVIVED. It searched for
 * getElementById('bCut'); the bug that killed the frame used
 * getElementById(id) with the id coming from an array — invisible to a
 * regex, fatal at runtime.
 *
 * So this executes. A fake DOM knows only the ids that are really in the
 * markup and returns null for anything else, which is exactly what a
 * browser does. If a component reaches for something that is not there,
 * this throws here instead of in Baba's hands.
 *
 * Two failures this would have caught:
 *   v101  a label constant declared in the wrong component file
 *   v121  buttons removed, their ids left in a forEach
 * Both silently removed half a feature and neither failed a test.
 *
 *     node tests/gastest/test_components.js
 */

const fs = require('fs');
const path = require('path');

const ROOT = path.join(__dirname, '..', '..');
let passed = 0, failed = 0;

function check(name, cond, detail) {
  if (cond) { passed++; console.log('  ok   ' + name); }
  else { failed++; console.log('  FAIL ' + name + (detail ? '  — ' + detail : '')); }
}

function fakeDom(ids) {
  const made = {};
  const el = (id) => {
    if (made[id]) return made[id];
    const style = {};
    const node = {
      id, style, className: '', textContent: '', value: '', disabled: false,
      hidden: false, dataset: {},
      classList: { add() {}, remove() {}, toggle() {}, contains() { return false; } },
      addEventListener() {}, removeEventListener() {}, focus() {}, blur() {},
      select() {}, setSelectionRange() {}, click() {},
      appendChild() {}, removeChild() {}, remove() {},
      setAttribute() {}, getAttribute() { return null; },
      removeAttribute() {}, insertBefore() {}, contains() { return false; },
      closest() { return null; }, matches() { return false; },
      firstElementChild: null, parentElement: null, children: [],
      innerHTML: '', innerText: '',
      getBoundingClientRect: () => ({ x: 0, y: 0, width: 100, height: 20 }),
      querySelector: () => el(id + '_q'), querySelectorAll: () => [],
      scrollWidth: 100, clientWidth: 100, scrollHeight: 100,
      selectionStart: 0, selectionEnd: 0,
      set onclick(v) {}, get onclick() { return null; },
      // EVERY node answers getContext. A canvas is a normal element to
      // getElementById, and the two frames that draw a scope reach for
      // its context straight away — my first harness gave getContext
      // only to querySelector results, so both scopes threw and the
      // failure was mine, not theirs.
      getContext: () => new Proxy({}, { get: () => () => ({}) }),
      width: 300, height: 60,
    };
    made[id] = node;
    return node;
  };

  const doc = {
    getElementById: (id) => (ids.has(id) ? el(id) : null),
    querySelector: () => el('_qs'),
    querySelectorAll: () => [],
    createElement: () => el('_new'),
    addEventListener() {}, removeEventListener() {},
    body: el('_body'), documentElement: el('_html'),
    // A frame that asks for a canvas gets one that answers everything.
    hidden: false,
  };
  el('_qs').getContext = () => ctx();
  el('_new').getContext = () => ctx();
  function ctx() {
    return new Proxy({}, { get: () => () => {} });
  }
  return doc;
}

function run(name) {
  const file = path.join(ROOT, name, 'index.html');
  const src = fs.readFileSync(file, 'utf8');

  const ids = new Set();
  for (const m of src.matchAll(/id="([A-Za-z0-9_-]+)"/g)) ids.add(m[1]);

  const script = src.split('<script>')[1] ? src.split('<script>')[1].split('</script>')[0] : '';
  check(name + ': has a script', script.length > 0);

  const document = fakeDom(ids);
  const listeners = [];
  const window = {
    document,
    parent: { postMessage() {} },
    addEventListener(t, fn) { listeners.push([t, fn]); },
    removeEventListener() {},
    requestAnimationFrame() { return 1; },
    cancelAnimationFrame() {},
    setInterval() { return 1; }, clearInterval() {},
    setTimeout() { return 1; }, clearTimeout() {},
    AudioContext: function () { return new Proxy({}, { get: () => () => ({}) }); },
    navigator: { mediaDevices: { getUserMedia: async () => ({}) } },
    MediaRecorder: function () { return new Proxy({}, { get: () => () => {} }); },
    FileReader: function () { return { readAsDataURL() {}, result: '' }; },
    Blob: function () { return { size: 0, type: '' }; },
    location: { href: '' }, localStorage: { getItem() { return null; }, setItem() {}, removeItem() {} },
    Audio: function () { return new Proxy({}, { get: () => () => {} }); },
    Image: function () { return new Proxy({}, { get: () => () => {} }); },
    URL: { createObjectURL: () => 'blob:x', revokeObjectURL() {} },
    fetch: async () => ({ ok: true, json: async () => ({}) }),
    Uint8Array, Float32Array, Number, Math, Date, JSON, String, Array,
    Object, console,
  };
  window.window = window;

  // NAMED ACCESS ON WINDOW. A browser exposes every element with an id
  // as a global of that name, and cassette_frontend uses it — `bFile`
  // with no lookup at all. Emulating it is what makes this harness
  // realistic rather than merely strict; without it the deck failed
  // here while working perfectly in Baba's hands, which is the wrong
  // kind of red.
  // ONLY the ids the script uses BARE. Passing every id as a parameter
  // collides with a script's own `const pad = ...` when an element is
  // also called pad — a SyntaxError that says nothing about the app.
  // A name counts as bare if it appears without a getElementById around
  // it and is never declared.
  const declared = new Set(
    [...script.matchAll(/\b(?:var|let|const|function)\s+([A-Za-z_$][\w$]*)/g)]
      .map((m) => m[1]));
  const named = [...ids].filter((id) =>
    !declared.has(id) && new RegExp('\\b' + id + '\\b').test(script));
  const namedVals = named.map((id) => document.getElementById(id));

  const fn = new Function(...named, 'window', 'document', 'navigator', 'MediaRecorder',
                          'FileReader', 'Blob', 'AudioContext', 'localStorage',
                          'requestAnimationFrame', 'cancelAnimationFrame',
                          'setInterval', 'clearInterval', 'setTimeout',
                          'clearTimeout', 'location', 'Audio', 'Image', 'URL',
                          'fetch',
                          '"use strict";' + script);
  try {
    fn(...namedVals, window, document, window.navigator, window.MediaRecorder,
       window.FileReader, window.Blob, window.AudioContext, window.localStorage,
       window.requestAnimationFrame, window.cancelAnimationFrame,
       window.setInterval, window.clearInterval, window.setTimeout,
       window.clearTimeout, window.location, window.Audio, window.Image,
       window.URL, window.fetch);
    check(name + ': THE SCRIPT RUNS without reaching for anything missing', true);
  } catch (e) {
    check(name + ': THE SCRIPT RUNS without reaching for anything missing',
          false, e.message);
    return;
  }

  // And a render message must not throw either — that is where labels
  // and arguments arrive, and where v121's dead lookups were waiting.
  const render = { data: { type: 'streamlit:render', args: {
    text: 'proba', scale: 1.0, source: 'mic',
    labels: { rec: 'rec', stop: 'stop', pause: 'pause',
              cut: 'cut', line: 'line' },
    sources: [{ id: 'mic', label: 'microphone' }],
  } } };
  try {
    for (const [type, fn2] of listeners) if (type === 'message') fn2(render);
    check(name + ': a render message does not throw', true);
  } catch (e) {
    check(name + ': a render message does not throw', false, e.message);
  }
}

console.log('COMPONENTS, EXECUTED\n');
for (const d of fs.readdirSync(ROOT)) {
  if (d.endsWith('_frontend') && fs.existsSync(path.join(ROOT, d, 'index.html'))) {
    run(d);
  }
}

console.log('\n' + passed + ' passed, ' + failed + ' failed');
process.exit(failed ? 1 : 0);

File: tests/test_components.py
"""NO COMPONENT MAY LOOK UP AN ELEMENT THAT IS NOT THERE.

This has killed a frame twice.

v101: a label constant was declared in the WRONG component, so
setRecording() threw "L_STOP is not defined" on the first press and took
the meter, the clock and the message line with it — while the button
still toggled, which is why it looked like it worked.

v121: the cut and line BUTTONS were removed and their ids left in a
forEach, so getElementById returned null, addEventListener threw, the
script died before ready() was ever sent, and Streamlit showed "trouble
loading the app.ttt_note component". The editor was simply absent and
pressing rec did nothing, because nothing was listening.

Both are the same shape: markup and script drifting apart, with no test
between them. A JavaScript exception fails nothing on its own — it just
quietly removes half the feature.

    python3 tests/test_components.py
"""

import os
import re
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

passed = failed = 0


def check(name, cond, detail=""):
    global passed, failed
    if cond:
        passed += 1
        print("  ok   " + name)
    else:
        failed += 1
        print("  FAIL " + name + ("  — " + str(detail) if detail else ""))


ROOT = os.path.join(os.path.dirname(__file__), "..")
FRONTENDS = [d for d in sorted(os.listdir(ROOT))
             if d.endswith("_frontend")
             and os.path.isfile(os.path.join(ROOT, d, "index.html"))]

print("EVERY COMPONENT'S SCRIPT AGAINST ITS OWN MARKUP\n")

check("1 the components are found", len(FRONTENDS) >= 3, FRONTENDS)

for name in FRONTENDS:
    src = open(os.path.join(ROOT, name, "index.html"), encoding="utf-8").read()

    ids = set(re.findall(r'id="([A-Za-z0-9_-]+)"', src))
    looked = set(re.findall(r"getElementById\(['\"]([A-Za-z0-9_-]+)['\"]\)", src))
    missing = sorted(looked - ids)
    check("2 %s looks up nothing that is missing from its markup" % name,
          not missing, missing)

    # A stray console.log is debug scaffolding that got shipped. v121
    # found three of them still in note_frontend from the v101 hunt.
    logs = len(re.findall(r"console\.log\(", src))
    check("3 %s ships no debug logging" % name, logs == 0, logs)

    # Every frame must announce itself, or Streamlit waits forever and
    # then reports that it cannot load the component.
    check("4 %s tells Streamlit it is ready" % name,
          "componentReady" in src, "no componentReady")

    # The script must at least parse. A syntax error is the same class of
    # failure: nothing renders and the app says only that it cannot load.
    script = src.split("<script>", 1)[-1].split("</script>", 1)[0]
    check("5 %s has balanced braces in its script" % name,
          script.count("{") == script.count("}"),
          "%d open, %d close" % (script.count("{"), script.count("}")))

print("\n{} passed, {} failed".format(passed, failed))

if __name__ == "__main__":
    sys.exit(1 if failed else 0)


def test_components():
    assert failed == 0, "%d of %d checks failed" % (failed, passed + failed)
